script/script.go

```diff
@@ -1,20 +1,96 @@
 package script
 
 import (
-	"bytes"
+	"context"
+	"fmt"
 	"io"
+	"io/ioutil"
+
+	"go.starlark.net/starlark"
 )
 
+// EntryPoint defines the name of the method that will be called when running a script (if it exists)
+// in addition to executing top level statements.
+const EntryPoint = "main"
+
 // A Script provides metadata and instructions to be executed by the interpreter.
 type Script struct {
-	io.Reader        // Read() instructions to execute
-	ID        string // Describes the ID of the script
+	io.Reader // Read() instructions to execute
+	io.Writer // Write() execution output
+	ID        string
+	Builtins  map[string]Func
+	Libraries map[string]Library
+}
+
+// Exec parses input from the script reader and executes it. It will also invoke the EntryPoint
+// method if one is available.
+func (script Script) Exec(ctx context.Context) error {
+	builtins := script.compilePredeclared()
+	thread := script.newThread()
+
+	symbols, err := starlark.ExecFile(thread, script.ID, script.Reader, builtins)
+	if err != nil {
+		// TODO: Better error type
+		return err
+	}
+
+	fn, ok := symbols[EntryPoint].(starlark.Callable)
+	if !ok {
+		return nil
+	}
+
+	res, err := starlark.Call(thread, fn, starlark.Tuple{}, []starlark.Tuple{})
+	if err != nil {
+		// TODO: Better error type
+		return err
+	}
+
+	if _, ok := res.(starlark.NoneType); !ok {
+		thread.Print(thread, res.String())
+	}
+
+	return nil
+}
+
+func (script Script) compilePredeclared() (builtins starlark.StringDict) {
+	for name, fn := range script.Builtins {
+		builtins[name] = fn.builtin(name)
```

Comment:
**[golangci-lint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
nilness: nil dereference in map update (govet)